### ISSUE TITLE
Document how to use R7RS

### ISSUE THREAD

**Issue by lassik** [gambit/gambit#408]
How do you make basic use of R7RS Scheme in gsi and gsc? I tried to find some documentation or examples on the subject but didn't come up with anything in the manual, wiki or mailing list.

**Comment by lassik** [comment]
For example, how do you compile the following R7RS program: https://github.com/srfi-explorations/missing-links This is a good example because it's simple and quite self-contained, but nevertheless relies on a fairly complex bundled library (chibi html-parser) where the difference between fast and slow builds is obvious.

**Comment by lassik** [comment]
Despite its name, (chibi html-parser) should be fully portable R7RS. I was able to run it in with no changes in Gauche and Cyclone in addition to Chibi. I'd love to add Gambit to the list.

**Comment by lassik** [comment]
According to the [Scheme Spheres R7RS repo](https://github.com/fourthbit/spheres) Gambit is currently missing `syntax-rules` macros and an R7RS-compatible library system, i.e. `define-library` and `import`. Are there plans to implement these in the near future?

**Comment by feeley** [comment]
No... there are plans to *release* them in the near future! ;-)

**Comment by lassik** [comment]
Awesome news, welcome to the R7RS family and thanks a lot for your tireless efforts with Gambit :)

If you need extra helping hands to test any R7RS stuff, I'm available.

**Comment by lassik** [comment]
There's a very good chance Gambit will instantly be the fastest R7RS implementation. Gerbil is already the fastest R7RS here: https://ecraven.github.io/r7rs-benchmarks/

**Comment by feeley** [comment]
Very doubtful Gambit will ever be faster than gerbil given that Gambit is gerbil's compiler...  I guess it can be faster momentarily when a new commit improves Gambit, until gerbil is rebased on the new commit.

**Comment by ColourGrey** [comment]
@feeley: If I understand your comment above correctly, 2 years ago Gambit was quite close to having a full implementation of the R7RS. On the other hand, the main page of the Gambit wiki says that _"The Gambit-C system conforms to the R4RS, R5RS and IEEE Scheme standards and implements all optional features."_ Could you please give a short clarifying follow-up about the progress on this front as of today? Thank you.

**Comment by amirouche** [comment]
Here is two examples:

- https://github.com/amirouche/gambit-scheme-pawns#how-to-compile-a-program-to-a-binary
- https://github.com/amirouche/gambit-scheme-pawns#how-to-create-a-library

**Comment by ColourGrey** [comment]
@amirouche: I don't see how that answers my question (and the one that started this discussion).

**Comment by feeley** [comment]
I have a rather literal view on the subject of "conformance", so currently Gambit does not conform to R7RS because there are a few missing features.  The most important one is the lack of hygienic macros.  If you disregard that aspect, Gambit is probably one of the most complete R7RS Scheme implementations available.

**Comment by pclouds** [comment]
Is there a checklist or something to keep track of these missing features? github issue could work for taht as well because it suports checkboxes in markdown.

**Comment by lassik** [comment]
IMHO grouping GitHub issues into [milestones](https://github.com/gambit/gambit/milestones) is a simple and effective way to do it.

**Comment by feeley** [comment]
Ideally some R7RS conformance test should be run to see what is still missing.  I quickly looked into the R7RS conformance test distributed with chibi but I disagree with some of the expected results.  For example it expects `(log 4096 2)` to return `12.0` but Gambit returns the exact result `12`.  Similarly it expects `(string-downcase "İ")` to return `"i̇"` but Gambit returns `"i"` (which I think is the correct result).  It would be good to develop a definitive (correct) R7RS conformance test using the chibi one as a starting point.

The hygienic macro support is probably Gambit's most important non conformance to R7RS.  This is being worked on but still far from a release.

**Comment by pclouds** [comment]
> I quickly looked into the R7RS conformance test distributed with chibi but I disagree with some of the expected results. For example it expects (log 4096 2) to return 12.0 but Gambit returns the exact result 12. Similarly it expects (string-downcase "İ") to return "i̇" but Gambit returns "i" (which I think is the correct result).

Tagging @ashinn jus in case the test suite in chibi could be updated with this (if he agrees of course)

**Comment by ashinn** [comment]
The R7RS test suite uses (chibi test), for which if we expect an inexact value, an exact result = to that value is accepted.  So in general, to better support multiple implementations, we should expect the inexact value.  Chibi will properly preserve exactness for some corner cases (e.g. `(sqrt 9)`), but not yet for the log case.

Gambit is actually returning the wrong result for `(string-downcase "İ")`.  The corresponding line from SpecialCasing.txt is:

`0130; 0069 0307; 0130; 0130; # LATIN CAPITAL LETTER I WITH DOT ABOVE`

where 0307 is the COMBINING DOT ABOVE.  The reason _why_ this redundant combining character is included is for round-trip compatibility - absent any locale, `(string-upcase "i")` would become `"I"`, losing the fact that this originally came from `"İ"`.

**Comment by feeley** [comment]
Gambit retuns the correct result for `(string-foldcase "İ")` => `"i̇"`.  It was my understanding of the R7RS spec that `(string-downcase s)` was equivalent to `(string-map char-downcase s)` and that this justified the existence of `string-foldcase`.  If my interpretation of the spec is incorrect it would be good to add an erratum or clarification.

The more general point I'm making is that the chibi R7RS conformance test is being overly constraining on the expected behaviour and some work would be required to fix the tests so that they accept all valid implementations of the R7RS spec.

**Comment by lassik** [comment]
> Gambit retuns the correct result for `(string-foldcase "İ")` => `"i̇"`. It is my understanding of the R7RS spec that `(string-downcase s)` is (or can be) equivalent to `(string-map char-downcase s)` and that this justified the existence of `string-foldcase`. If my interpretation of the spec is incorrect it would be good to add an erratum or clarification.

@johnwcowan Can you clarify this?

**Comment by ashinn** [comment]
The standard is fairly clear on this point, in the description of `string-*case`:

> These procedures apply the Unicode full string uppercas-
> ing, lowercasing, and case-folding algorithms to their ar-
> guments and return the result.

The term "full" here has a formal meaning in the Unicode standard with respect to case mappings, and specifically means that you can't consider case mappings on an individual character level (i.e. the number of codepoints can change).  A special exception was made only for the special case handling of Σ, which is allowed to be optionally supported due to the complexity of determining word breaks.

Due to the implementation burden it may have been preferable to make the full mappings only recommended, not required (though you could always just drop all Unicode support and thus be conforming).  Generally though, in a Unicode world, the idea of strings as fixed arrays of codepoints which can be processed in isolation is broken, and this was one of the complaints against R6RS.  We felt strongly enough about this that some wanted to go further and remove the individual character mappings `char-*case` specifically to avoid their mis-use (as happened in Gambit).

So the current test case for `(string-downcase "İ")` is the only conforming result, whereas `(log 4096 2)` is allowed to return either 12 or 12.0, and the test does allow either.

If there are other cases where the test suite is too demanding we should make them more lenient and/or flag guarded (it's nice to be able to track maximal conformance as well as minimal).

**Comment by johnwcowan** [comment]
> The term "full" here has a formal meaning in the Unicode standard with respect to case mappings, and specifically means that you can't consider case mappings on an individual character level (i.e. the number of codepoints can change).

Exactly so.

> A special exception was made only for the special case handling of Σ, which is allowed to be optionally supported due to the complexity of determining word breaks.

That was part of the reason.  Unicode only requires detecting a word break following "Σ" and downcasing to "ς" rather than "σ".  But in working with actual Greek, to downcase the string "ΦΊΛΟΣ." you have  to know whether this is "φίλος" (the Ancient Greek word for 'friend') at the end of a sentence or the Modern Greek "φίλοσ.", an abbreviation for φιλοσοφία 'philosophy'.  So Unicode only gets the common case right anyhow, not all cases.

**Comment by nat-418** [comment]
`gsc -:r7rs `and `gsi -:r7rs`

**Comment by jhidding** [comment]
> For anyone who finds this thread through a search engine: Gambit supports R7RS via `gsc -:r7rs `and `gsi -:r7rs` flags.

Not having hygienic macro support (and not even mentioning that explicitly in the documentation) is a very big omission though. Most r7rs code will not run correctly.
